Content.Client/UserInterface/CooldownGraphic.cs

```diff
@@ -8,28 +8,49 @@ namespace Robust.Client.UserInterface.Controls
 {
     public class CooldownGraphic : Control
     {
-        public float Fraction { get; set; }
+        /// <summary>
+        ///     Progress of the cooldown animation.
+        ///     Possible values range from 1 to -1, where 1 to 0 is a depleting circle animation and 0 to -1 is a blink animation.
+        /// </summary>
+        public float Progress { get; set; }
 
         protected override void Draw(DrawingHandleScreen handle)
         {
             const int maxSegments = 64;
             const float segment = MathHelper.TwoPi / maxSegments;
 
-            var segments = (int)Math.Max(2, Math.Ceiling(maxSegments * Fraction)); // ensure that we always have at least 3 vertices
-            var max = MathHelper.TwoPi * Fraction;
-            var radius = (Math.Min(SizeBox.Height, SizeBox.Width) / 2) * 0.875f; // 28/32 = 0.875 - 2 pixels inwards from the edge
-            var ring_radius = (Math.Min(SizeBox.Height, SizeBox.Width) / 2) * 0.5625f;
+            var effective_fraction = Progress >= 0f ? Progress : 1f; // we want the full circle for the blink
+
+            var segments = (int)Math.Max(2, Math.Ceiling(maxSegments * effective_fraction) + 1); // ensure that we always have at least 3 vertices, we also need one more segment
+            var max = MathHelper.TwoPi * effective_fraction;
+            var outer_radius = (Math.Min(SizeBox.Height, SizeBox.Width) / 2) * 0.875f; // 28/32 = 0.875 - 2 pixels inwards from the edge
+            var inner_radius = (Math.Min(SizeBox.Height, SizeBox.Width) / 2) * 0.5625f; // 18/32 = 0.5625 - 5 pixels thick
 
             Span <Vector2> vertices = stackalloc Vector2[segments * 2];
```

Comment:
Wait this really stack-allocates this whole thing? What is that at maximum, like 1kb of stack?

Reply:
Yeah 1kB + span's ptr, offset and len (16)

Reply:
Looks like the default stack is 1mB, so it's not as bad as I first thought.

Reply:
Oh you meant the how big can the stack get...